tar/private/tar.bzl

```diff
@@ -427,29 +437,41 @@ def _vis_encode(filename):
     # TODO(#794): correctly encode all filenames by using vis(3) (or porting it)
     return filename.replace(" ", "\\040")
 
-def _expand(file, expander, transform = to_repository_relative_path):
-    expanded = expander.expand(file)
+def _add_parent_lines(lines, path, seen_parents):
+    parent_lines = []
+    for i in range(len(path) - 1, 0, -1):
+        if path[i] == "/":
+            parent = path[:i + 1]
+            if parent in seen_parents:
+                break
+            seen_parents[parent] = True
+            parent_lines.append(_mtree_line(_vis_encode(parent), "dir"))
+    lines.extend(reversed(parent_lines))
+
+def _expand(file, expander, path = None):
+    if not path:
+        path = to_repository_relative_path(file)
     lines = []
-    for e in expanded:
-        path = transform(e)
-        segments = path.split("/")
-        for i in range(1, len(segments)):
-            parent = "/".join(segments[:i])
-
-            # NOTE: The mtree format treats file paths without slashes as "relative" entries.
-            #       If a relative entry is a directory, then it will "change directory" to that
-            #       directory, and any subsequent "relative" entries will be created inside that
-            #       directory. This causes issues when there is a top-level directory that is
-            #       followed by a top-level file, as the file will be created inside the directory.
-            #       To avoid this, we append a slash to the directory path to make it a "full" entry.
-            if i == 1:
-                parent += "/"
-
-            lines.append(_mtree_line(_vis_encode(parent), "dir"))
-
-        lines.append(_mtree_line(_vis_encode(path), "file", content = _vis_encode(e.path)))
+
+    # The result of this callback is uniquified, so seen_parents is purely an optimization for large
+    # directories.
+    seen_parents = {}
+    if not file.is_directory:
+        _add_parent_lines(lines, path, seen_parents)
+        lines.append(_mtree_line(_vis_encode(path), "file", content = _vis_encode(file.path)))
+    else:
+        for e in expander.expand(file):
+            child_path = path + "/" + e.tree_relative_path
+            _add_parent_lines(lines, child_path, seen_parents)
+            lines.append(_mtree_line(_vis_encode(child_path), "file", content = _vis_encode(e.path)))
     return lines
 
+def _expand_root_symlink(symlink, expander):
+    return _expand(symlink.target_file, expander, symlink.path)
```

Comment:
if i am not misunderstanding the logic here, this is basically dereferencing symlink target into original symlink location.  my worry with this is that its gonna increase the size of archives generated by tar, ideally we'd like to preserve symlinks (when preserve_symlinks = True). 

Maybe I am  missing something here,  what's the reasoning for not preserving these smylinks?

Reply:
The runfiles symlinks are not symlinks in the strict sense any more than all the regular runfiles, they are just staged at a custom path rather than the rlocationpath. That's why I wouldn't treat them differently.

Reply:
Ah i see, that makes sense.  Do you think this will be a deoptimization where target_file might also be in the archive and we are dereferencing by default instead of symlinking? 

I also don't know if Bazel's runfiles implementation deferences these during runfiles tree generation.